### PR TITLE
BUG fix for date_range 56134

### PR DESCRIPTION
**Pull request by jrmylow**
- [X] closes https://github.com/pandas-dev/pandas/issues/56134
- [X] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [X] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [X] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.
- [X] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

Edit, the first commit is incorrectly named - it fixes 56134 instead

### CODE REVIEW

**Comment by jrmylow** [comment]
pre-commit.ci autofix

File: doc/source/whatsnew/v2.3.0.rst [hunk 1]
```diff
@@ -209,6 +209,7 @@ Styler
 
 Other
 ^^^^^
+- Bug in :func:`date_range` where the last valid timestamp would sometimes not be produced (:issue:`56134`)
```
Comment (MarcoGorelli) [inline doc/source/whatsnew/v2.3.0.rst:212]: there's a `Datetimelike` section, maybe this can go there?

File: pandas/core/arrays/datetimes.py [hunk 2]
```diff
@@ -2770,16 +2770,12 @@ def _generate_range(
     else:
         end = None
 
+    # GH #56134 fix bug in end of interval
```
Comment (MarcoGorelli) [inline pandas/core/arrays/datetimes.py:2773]: I don't think this is necessary

**Review by MarcoGorelli — COMMENTED** [review]
nice!

**Comment by MarcoGorelli** [comment]
Thanks for updating!

> Edit, the first commit is incorrectly named - it fixes 56134 instead

it's fine, commits are squashed together anyway, the title of the PR is more important

**Review by MarcoGorelli — APPROVED** [review]
This is the best kind of bug fix - just, delete code, and the issue goes away

Looks good to me on green

Awesome stuff here @jrmylow !

**Pull request merged**